File: array/sliding_window/sliding_window3_lc209_star.py


# 不定长的滑动窗口
def minSubArrarLen(s, nums):
    """
        长度最小的子数组
        描述: 找出该数组中满足其和>=target的长度最小的连续子数组[numsl, numsl+1, ..., numsr-1, numsr],
        并返回其长度. 如果不存在符合条件的子数组, 返回0
    """
    start = 0
    cur_sum = 0
    min_len = float('inf')
    for end in range(len(nums)):
        cur_sum += nums[end]
        while cur_sum >= s:             # 有的题用if, 这边为什么用while, 想想就知道了其实
            min_len = min(min_len, end - start + 1)     # 窗口长度end-start+1
            cur_sum -= nums[start]      # 这步跟下一步就是用来移动窗口的, 窗口改变, 相应的值也要减去
            start += 1                  # 移动窗口, 只要移动了窗口, cur_sum就要变, 就是说要有上一步
        # 出了while循环, 说明cur_sum < s, 这个时候就继续进行for循环移动end把窗口变大就可以了
    
    if min_len == float('inf'):         # 如果根本没变化就说明找遍了都不符合条件, 根据要求返回0
        return 0
    else:
        return min_len

aaa = minSubArrarLen(s=7, nums=[2,3,1,2,4,3])
print(aaa)

File: cpp/array/sliding_window/sliding_window3_lc209_star.cpp
/*
    字符串的排列
    给你两个字符串 s1 和 s2 ，写一个函数来判断 s2 是否包含 s1 的排列。如果是，返回 true ；否则，返回 false 。
    换句话说，s1 的排列之一是 s2 的 子串 。
*/

#include <iostream>
#include <vector>
using namespace std;

int minSubArrayLen(int target, vector<int> &nums)
{
    int start = 0;
    int cur_sum = 0;
    int min_len = INT_MAX;
    for (int end = 0; end < nums.size(); end++)
    {
        cur_sum += nums[end];
        while (cur_sum >= target)
        {
            min_len = min(min_len, end - start + 1);
            cur_sum -= nums[start];
            start += 1;
        }
    }
    if (min_len == INT_MAX)
    {
        return 0;
    }
    else
    {
        return min_len;
    }
}
int main()
{
    int target = 7;
    vector<int> nums = {2, 3, 1, 2, 4, 3};
    int result = minSubArrayLen(target, nums);
    cout << result << endl; // true
    return 0;
}